datafusion/physical-expr-common/src/binary_view_map.rs

```diff
@@ -0,0 +1,581 @@
+
+                let len = *v as u32;
+                if len as usize != value.len() {
+                    return false;
+                }
+
+                // We should probably change arrow-rs to provide a
+                // GenericByteViewArray::value_from_view() method
+                let b = if len <= 12 {
+                    unsafe {
+                        GenericByteViewArray::<BinaryViewType>::inline_value(
+                            v,
+                            len as usize,
+                        )
+                    }
+                } else {
+                    let view = ByteView::from(*v);
+
+                    let data = unsafe {
+                        self.view_buffers.get_unchecked(view.buffer_index as usize)
+                    };
+                    let offset = view.offset as usize;
+                    unsafe { data.get_unchecked(offset..offset + len as usize) }
+                };
+                b == value
+            });
+
+            let payload = if let Some(entry) = entry {
+                entry.payload
+            } else {
+                // no existing value, make a new one.
+                let payload = make_payload_fn(Some(value));
+                let new_header = Entry { view_idx, payload };
+
+                let view = if value.len() <= 12 {
+                    unsafe { *values.views().get_unchecked(view_idx) }
+                } else {
+                    let v = unsafe { *values.views().get_unchecked(view_idx) };
+                    let mut v = ByteView::from(v);
+                    v.buffer_index += buffer_offset as u32;
+                    v.as_u128()
+                };
+                self.views.push(view);
+                self.map
+                    .insert_accounted(new_header, |_| hash, &mut self.map_size);
+                payload
+            };
+            observe_payload_fn(payload);
+        }
+    }
+
+    /// Converts this set into a `StringViewArray`, or `BinaryViewArray`,
+    /// containing each distinct value
+    /// that was inserted. This is done without copying the values.
+    ///
+    /// The values are guaranteed to be returned in the same order in which
+    /// they were first seen.
+    pub fn into_state(self) -> ArrayRef {
+        // Only make a `NullBuffer` if there was a null value
+        let nulls = self.null.map(|(_payload, null_index)| {
+            let num_values = self.views.len();
+            single_null_buffer(num_values, null_index)
+        });
+
+        let len = self.views.len();
+        let b = ScalarBuffer::new(Buffer::from_vec(self.views), 0, len);
+        match self.output_type {
+            OutputType::BinaryView => {
+                // SAFETY: the offsets were constructed correctly
+                let mut array = unsafe {
+                    BinaryViewArray::new_unchecked(b, self.view_buffers, nulls)
+                };
+                if array.len() < (self.input_len / 2) {
```

Comment:
We need to gc the underlying buffer if the cardinality shrinks, this prevents we holding the buffer for too long and cause excessive memory usage.

The current heuristic is that if the array length cut by more than half, we should gc it. Let me know if there's a better strategy.